pika/adapters/twisted_connection.py

```diff
@@ -297,82 +218,75 @@ class TwistedConnection(base_connection.BaseConnection):
-            # docstrings for why it's OK to pass None as the file descriptor)
-            self.ioloop.update_handler(None, self.event_state)
-        return error
-
-    def _adapter_disconnect(self):
-        """Called when the adapter should disconnect"""
-        self.ioloop.remove_handler(None)
-        self._cleanup_socket()
-
-    def _on_connected(self):
-        """Call superclass and then update the event state to flush the outgoing
-        frame out. Commit 50d842526d9f12d32ad9f3c4910ef60b8c301f59 removed a
-        self._flush_outbound call that was in _send_frame which previously
-        made this step unnecessary.
+            nbio=nbio,
+            internal_connection_workflow=internal_connection_workflow)
+
+    @classmethod
+    def create_connection(cls,
+                          connection_configs,
+                          on_done,
+                          custom_ioloop=None,
+                          workflow=None):
+        """Implement
+        :py:classmethod:`pika.adapters.BaseConnection.create_connection()`.
 
         """
-        super(TwistedConnection, self)._on_connected()
-        self._manage_event_state()
+        nbio = _TwistedIOServicesAdapter(custom_ioloop)
+
+        def connection_factory(params):
+            """Connection factory."""
+            if params is None:
+                raise ValueError('Expected pika.connection.Parameters '
+                                 'instance, but got None in params arg.')
+            return cls(
+                parameters=params,
+                custom_ioloop=nbio,
+                internal_connection_workflow=False)
+
+        return cls._start_connection_workflow(
+            connection_configs=connection_configs,
+            connection_factory=connection_factory,
+            nbio=nbio,
+            workflow=workflow,
+            on_done=on_done)
 
     def channel(self, channel_number=None):
         """Return a Deferred that fires with an instance of a wrapper around the
         Pika Channel class.
 
         """
         d = defer.Deferred()
-        base_connection.BaseConnection.channel(self, channel_number, d.callback)
+        super(TwistedConnection, self).channel(channel_number, d.callback)
         return d.addCallback(TwistedChannel)
 
-    # IReadWriteDescriptor methods
-
-    def fileno(self):
-        return self.socket.fileno()
-
-    def logPrefix(self):
-        return "twisted-pika"
-
-    def connectionLost(self, reason):
-        # If the connection was not closed cleanly, log the error
-        if not reason.check(error.ConnectionDone):
-            log.err(reason)
-
-        self._on_terminate(connection.InternalCloseReasons.SOCKET_ERROR,
-                           str(reason))
-
-    def doRead(self):
-        self._handle_read()
-
-    def doWrite(self):
-        self._handle_write()
-        self._manage_event_state()
-
 
 class TwistedProtocolConnection(base_connection.BaseConnection):
```

Comment:
This PR implements a common protocol/transport model for all supported non-blocking adapters as well as non-blocking connection establishment. The `TwistedConnection` class above already benefits from this. Because of this and lack of tests, I would prefer to get rid of TwistedProtocolConnection. If can't remove TwistedProtocolConnection class, then someone would need to implement a test suite for it and debug.